perf(core): minor improvements to listener instructions

crisbeto opened the pull request:
Makes the following improvements to the listener instructions to make them slightly smaller and more memory-efficient.

1. Removes the default value from the `useCapture` parameter since it generates more code than just castint to `false`.
2. Removes the `useCapture` and `eventTargetResolver` parameters from `ɵɵsyntheticHostListener` since they won't be generated by the compiler, as far as I can tell.
3. Makes it so that we don't have to return a target name from a `GlobalTargetResolver`. This allows us to save on some memory, because we can return a reference to the target without having to wrap it in an object literal.

DEPRECATIONS:
`EventManagerPlugin.getGlobalEventTarget` is now deprecated and won't be called from Ivy code anymore. Global events will go through `addEventListener`.

AndrewKushnir commented on packages/core/src/render3/instructions/listener.ts:
```diff
@@ -176,11 +173,9 @@ function listenerInternal(
+        // The first argument of `listen` function in Procedural Renderer is an object reference.
         listenerFn = wrapListener(tNode, lView, context, listenerFn, false /** preventDefault */);
-        const cleanupFn = renderer.listen(resolved.name || target, eventName, listenerFn);
+        const cleanupFn = renderer.listen(target as RElement, eventName, listenerFn);
```
It looks like this change will result in a different codepath here (for DefaultDomRenderer2 implementation):
https://github.com/angular/angular/blob/99f2ffc740279509b89bb96093e2515ce0b46ae1/packages/platform-browser/src/dom/dom_renderer.ts#L263-L273
Which in turn will never call the `eventManager.addGlobalEventListener` function, thus potentially causing regressions.

crisbeto replied:
I noticed that, but eventually it gets mapped back to the global objects anyway. https://github.com/angular/angular/blob/295889ed2ee5d1382abc60a2023795ac3c02084b/packages/platform-browser/src/browser/browser_adapter.ts#L60

AndrewKushnir replied:
If I understand correctly, there might be some custom event manager plugins (added via `EVENT_MANAGER_PLUGINS` multi provider), which may implement the `getGlobalEventTarget` function and expect it to be called. If we avoid calling the `addGlobalEventListener` branch, the `addEventListener` function would be invoked on these plugin instances where previously we called the `addGlobalEventListener` one.

crisbeto replied:
That's true, but everything has to go through `addEventListener` eventually so I suspect that it won't be as breaking. We can run it through g3 to see if it breaks anything.

AndrewKushnir replied:
Yeah, I agree that it'd probably not be very break-y (may only affect custom event plugins that handle global events in a special way) and I've started a [presubmit](http://test/OCL:370789059:BASE:370789080:1619566741219:95e40c2) to see if there are other use-cases (and will need to run a global one too). I guess if we decide to go that route, we should probably consider deprecating `addGlobalEventListener` API first, WDYT? (btw, I'm curious what was the reason to introduce it originally)

crisbeto replied:
Thank you for the presubmits and I agree that we should deprecate `addGlobalEventListener`. I don't know the context behind it, but maybe it's something specific to ViewEngine?

AndrewKushnir replied:
Hey @crisbeto, global presubmit looks good too! :)

I've found a PR that introduces the `addGlobalEventListener` support: #1255. Based on https://github.com/angular/angular/pull/1255#discussion-diff-27938552, my understanding is that it was needed to workaround an issue in Dart, related to zones.

The change in this PR makes sense to me, I just propose to treat it as a breaking one (thus land in next major - v13). WDYT?

crisbeto replied:
Sure, that makes sense. Should we bring it up in the next sync?

AndrewKushnir replied:
> Should we bring it up in the next sync?

@crisbeto yep, added to the meeting agenda.

crisbeto commented:
I've made the changes that we discussed during the sync meeting.

AndrewKushnir commented on packages/platform-browser/src/browser/browser_adapter.ts:
```diff
@@ -57,6 +57,7 @@ export class BrowserDomAdapter extends GenericBrowserDomAdapter {
     return node instanceof DocumentFragment;
   }
 
+  /** @deprecated No longer being used. To be removed in a future version. */
```
nit: since VE code still uses that, we can just mention that it's unused in Ivy here (and in a couple other places updated in this PR):
```suggestion
  /** @deprecated No longer being used in Ivy code. To be removed in a future version. */
```

AndrewKushnir commented on packages/core/src/render3/instructions/listener.ts:
```diff
@@ -176,11 +173,9 @@ function listenerInternal(
-        // The first argument of `listen` function in Procedural Renderer is:
-        // - either a target name (as a string) in case of global target (window, document, body)
-        // - or element reference (in all other cases)
+        // The first argument of `listen` function in Procedural Renderer is an object reference.
```
I think this comment is no longer needed, since there is just one type of argument and the code is quite clear on what that argument is.
```suggestion
```

AndrewKushnir reviewed (approved):
LGTM (just a couple small comments), thanks for the update @crisbeto 👍 

Btw, may be we can rename the commit to `perf(core): ...`, WDYT?

I think we should also include a deprecation note to the commit message (so that the changelog generator can pick it up), [see an example here](https://github.com/angular/angular/commit/e148382bd0c6ba5c442fd6fa8c95f957205b5d01).

crisbeto commented:
The latest set of feedback has been addressed.

AndrewKushnir commented:
@crisbeto thanks for addressing the feedback. What do you think about updating the commit message to become `perf(core): ...` and also including deprecation note (as mentioned in https://github.com/angular/angular/pull/41807#pullrequestreview-648477507)?

AndrewKushnir commented:
[Presubmit](http://test/OCL:370789059:BASE:371176890:1619724238837:8ca606f2).

crisbeto committed:
perf(core): minor improvements to listener instructions

Makes the following improvements to the listener instructions to make them slightly smaller and more memory-efficient.

1. Removes the default value from the `useCapture` parameter since it generates more code than just castint to `false`.
2. Removes the `useCapture` and `eventTargetResolver` parameters from `ɵɵsyntheticHostListener` since they won't be generated by the compiler, as far as I can tell.
3. Makes it so that we don't have to return a target name from a `GlobalTargetResolver`. This allows us to save on some memory, because we can return a reference to the target without having to wrap it in an object literal.

DEPRECATIONS:
`EventManagerPlugin.getGlobalEventTarget` is now deprecated and won't be called from Ivy code anymore. Global events will go through `addEventListener`.

crisbeto commented:
Done, although I'm not 100% sure that this is the correct way to format the deprecations in order for them to show up in the changelog.

AndrewKushnir commented:
[Presubmit](http://test/OCL:370789059:BASE:371191894:1619728256925:62bf98c2).

kyliau commented on packages/platform-browser/src/browser/browser_adapter.ts:
```diff
@@ -57,6 +57,7 @@ export class BrowserDomAdapter extends GenericBrowserDomAdapter {
     return node instanceof DocumentFragment;
   }
 
+  /** @deprecated No longer being used in Ivy code. To be removed in a future version. */
```
I suggest we make the deprecation more concrete and remove it in v14.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

Also FYI presubmits are successful for the changes in this PR.

crisbeto committed:
fixup! perf(core): minor improvements to listener instructions

crisbeto commented:
I pushed another commit to update the doc strings. I don't think that we need another presubmit for it.

mhevery reviewed (approved):
reviewed-for: global-approvers
